src/Temporalio/Bridge/CustomSlotSupplier.cs

```diff
@@ -59,108 +61,145 @@ private static Temporalio.Worker.Tuning.SlotInfo SlotInfoFromBridge(Interop.Temp
-                                    cancelTokenSrc.Token);
+                                result = Interop.Methods.temporal_core_complete_async_reserve(
+                                    (Interop.TemporalCoreSlotReserveCompletionCtx*)completionCtx.ToPointer(),
+                                    new(permitId.ToPointer()));
                             }
-                            var permit = await reserveTask.ConfigureAwait(false);
-                            unsafe
+
+                            if (result == 0)
                             {
-                                var usedPermitId = AddPermitToMap(permit);
-                                Interop.Methods.temporal_core_complete_async_reserve(sender.ToPointer(), new(usedPermitId));
+                                // We need to undo the reservation
+                                CompleteCancelReserve(completionCtx);
+                                Release(null, permitId);
                             }
-                            return;
+
+                            break;
                         }
                         catch (OperationCanceledException) when (cancelTokenSrc.Token.IsCancellationRequested)
                         {
-                            unsafe
-                            {
-                                // Always call this to ensure the sender is freed
-                                Interop.Methods.temporal_core_complete_async_reserve(sender.ToPointer(), new(0));
-                            }
-                            return;
+                            CompleteCancelReserve(completionCtx);
+                            break;
                         }
 #pragma warning disable CA1031 // We are ok catching all exceptions here
                         catch (Exception e)
                         {
 #pragma warning restore CA1031
                             logger.LogError(e, "Error reserving slot");
                         }
+
                         // Wait for a bit to avoid spamming errors
-                        await Task.Delay(1000, cancelTokenSrc.Token).ConfigureAwait(false);
+                        try
+                        {
+                            await Task.Delay(1000, cancelTokenSrc.Token).ConfigureAwait(false);
+                        }
+                        catch (OperationCanceledException)
+                        {
+                            CompleteCancelReserve(completionCtx);
+                            break;
+                        }
                     }
                 }
+#pragma warning disable CA1031 // The task is detached, logging here is the only way to observe exceptions.
+                catch (Exception e)
+                {
+#pragma warning restore CA1031
+                    logger.LogError(e, "Exception escaped reserve slot retry loop");
+                }
+                finally
+                {
+                    reservationCancelSources.TryRemove(completionCtx, out _);
```

Comment:
```suggestion
                    reservationCancelSources.Remove(completionCtx);
```

I think this is good enough IIUC

Reply:
`ConcurrentDictionary` has explicit implementation of `IDictionary`, calling `Remove` would require an ugly type cast.